fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ItemPredicateMixin.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.item;
+
+import java.util.List;
+import java.util.Optional;
+
+import com.mojang.datafixers.util.Either;
+import com.mojang.serialization.Codec;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Mutable;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.Unique;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import net.minecraft.item.ItemStack;
+import net.minecraft.predicate.NumberRange;
+import net.minecraft.predicate.item.ItemPredicate;
+import net.minecraft.util.dynamic.Codecs;
+
+import net.fabricmc.fabric.api.item.v1.CustomItemPredicate;
+import net.fabricmc.fabric.api.item.v1.FabricItemPredicate;
+import net.fabricmc.fabric.impl.item.ItemPredicateExtensions;
+import net.fabricmc.fabric.impl.item.VanillaItemPredicate;
+
+@Mixin(ItemPredicate.class)
+abstract class ItemPredicateMixin implements ItemPredicateExtensions, FabricItemPredicate {
+	@Shadow
+	@Final
+	@Mutable
+	public static Codec<ItemPredicate> CODEC;
+
+	@Unique
+	private List<CustomItemPredicate> custom = List.of();
+
+	@Override
+	public List<CustomItemPredicate> custom() {
+		return custom;
+	}
+
+	@Override
+	public void fabric_setCustom(List<CustomItemPredicate> custom) {
+		this.custom = custom;
+	}
+
+	@Inject(method = "<clinit>", at = @At("TAIL"))
+	private static void modifyCodec(CallbackInfo ci) {
+		Codec<ItemPredicate> vanillaCodec = CODEC;
+		VanillaItemPredicate.register(vanillaCodec);
+
+		Codec<CustomItemPredicate> customCodec = CustomItemPredicate.REGISTRY.getCodec().dispatch("fabric:type", CustomItemPredicate::getCodec, c -> c);
+
+		CODEC = Codecs.either(Codecs.nonEmptyList(customCodec.listOf()), vanillaCodec).xmap(e -> e.map(l -> {
```

Comment:
Is there a particular reason to require custom predicates to be part of a list instead of a single object like vanilla? Seems like boilerplate.

Reply:
AFAIK there's no way to dispatch codec using the parent key

Reply:
Couldn't a snippet like this do the trick? 
```java
Codecs.either(vanillaCodec, customCodec)
    .xmap(
        e -> e.map(l -> l, r -> r),
        predicate -> {
            CustomItemPredicate custom = predicate.custom(); // imagine an API like FabricIngredient
            if (custom == null) {
                return Either.left(predicate);
            } else {
                return Either.right(custom);
            }
        }
    )
```
That way symmetry is maintained between the presentation of each type on the JSON level, and avoids having to patch something into `ItemPredicate` itself.

Reply:
That makes custom and vanilla predicate can't be used at the same time.

Reply:
I see your point. It could be solved with more manipulation, but it's not strictly necessary. I just dislike that it's impossible to use a custom predicate without brackets.